utils/src/main/java/org/robolectric/util/TempDirectory.java

```diff
@@ -24,13 +39,26 @@ public TempDirectory(String name) {
       throw new RuntimeException(e);
     }
 
-    // Use a manual hook that actually clears the directory
-    // This is necessary because File.deleteOnExit won't delete non empty directories
-    Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
-      @Override public void run() {
-        destroy();
+    synchronized (TempDirectory.class) {
+      // If we haven't initialised the shutdown hook we should set everything up.
+      if (tempDirectoriesToDelete == null) {
```

Comment:
nit: would it be preferable to initialize this upfront like this:

```java
private static Set<TempDirectory> tempDirectoriesToDelete = Collections.synchronizedSet(new HashSet<>());
```

And then do:

```java
synchronized (tempDirectoriesToDelete) {
    if (tempDirectoriesToDelete.size() == 0) { ...}
}
```

Reply:
Done.